### PR TITLE
TKR api conversion webhooks.

### LINKED ISSUE #1849: TKR API conversion webhook
Implement conversion webhooks for the TKR APIs.
- Should be able to provide lossless conversion between v1alpha1 and v1alpha3 TKR APIs.

**Affected product area (please put an X in all that apply)**

- [x] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by ShashankGirish**
### What this PR does / why we need it

Implements the TKR API conversion webhooks.
- Makes tkr v1alpha3 api the hub type.
- Add conversion logic in v1lpha1 to the hub type and unit tests for this.
- Generate webhooks for the `TanzuKubernetesRelease` types.

Also:
- Introduces the conversion-gen module to generate some of the conversion code.
- Updates the `PROJECT` file and generates/fills some incomplete rbac and sample yamls related to TKR apis.
- Uncomments the webhook and certmanager sections in kustomization.yaml.
- Adds  var-naming, ST1003, receiver-naming, and ST1016 lint exclusions for conversion code.
    - The generated conversion stubs have underscores in function names, thus these require excluding.
    - The conversion code is easier to read when referred to as src/dest or spoke/hub.

### Which issue(s) this PR fixes

Fixes https://github.com/vmware-tanzu/tanzu-framework/issues/1849

### Describe testing done for PR

- Unit testing.
- Manually deployed on AWS and tested conversion.

### Release note

```release-note

```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/go.mod [hunk 1]
```diff
@@ -78,10 +78,11 @@ require (
...
-	k8s.io/component-base v0.23.4 // indirect
-	k8s.io/kube-openapi v0.0.0-20220124234850-424119656bbf // indirect
-	sigs.k8s.io/json v0.0.0-20211208200746-9f7c6b3444d2 // indirect
-	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
+	k8s.io/cluster-bootstrap v0.23.0 // indirect
+	k8s.io/component-base v0.23.0 // indirect
+	k8s.io/kube-openapi v0.0.0-20211115234752-e816edb12b65 // indirect
+	k8s.io/kubectl v0.23.0 // indirect
+	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.0 // indirect
```
[the remark this answers is not in the document]
Comment (ShashankGirish) [reply addons/go.mod:86]: This is unintentional. I'll fix this.

[the remark this answers is not in the document]
Comment (ShashankGirish) [reply addons/go.mod]: Should be fixed now.

File: apis/run/v1alpha1/conversion.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+	// Transform the container images.
+	for index, image := range in.Images {
+		switch in.Images[index].Name {
+		case etcdContainerImageName:
+			out.Kubernetes.Etcd = &v1alpha3.ContainerImageInfo{
+				ImageRepository: image.Repository,
+				ImageTag:        image.Tag,
+			}
+		case pauseContainerImageName:
```
Comment (prkalle) [inline apis/run/v1alpha1/conversion.go:91]: I see in the current v1alpha1 TKR there are a bunch of images available for `pause` image (as shown below). Hope there isn't any issue copying the last entry of `pause` image in the v1alpha1 TKR list to v1alpha3 TKR's `pause` image
```
 - name: pause
    repository: projects-stg.registry.vmware.com/tkg
    tag: "3.5"
  - name: pause
    repository: projects-stg.registry.vmware.com/tkg
    tag: 3.5-windows-amd64
  - name: pause
    repository: projects-stg.registry.vmware.com/tkg
    tag: 3.5-windows-amd64-1903
```

File: apis/run/v1alpha3/tanzukubernetesrelease_webhook.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var tanzukubernetesreleaselog = logf.Log.WithName("tanzukubernetesrelease-resource")
+
+func (r *TanzuKubernetesRelease) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	// TODO: Remove logger call below, as it is only used to prevent deadcode lint errors.
+	tanzukubernetesreleaselog.Info("TODO: Remove this - this is only to pass lint")
```
Comment (prkalle) [inline apis/run/v1alpha3/tanzukubernetesrelease_webhook.go:16]: Can we remove this log or add some meaningful log?

Comment (ShashankGirish) [reply apis/run/v1alpha3/tanzukubernetesrelease_webhook.go:16]: Removed.

File: apis/run/v1alpha1/conversion_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	}))
+
+	// Add other types here in future
+}
+
+func TestHubSpokeHub(t *testing.T) {
+	t.Run("for TanzuKubernetesRelease", func(t *testing.T) {
+		hub := &v1alpha3.TanzuKubernetesRelease{
+			Spec: v1alpha3.TanzuKubernetesReleaseSpec{
+				Version: "#ŉƈOƕʘ賡谒湪ȥ#4",
```
[the remark this answers is not in the document]
Comment (ShashankGirish) [reply apis/run/v1alpha1/conversion_test.go:32]: Indeed, there are weird characters. That doesn't matter though, the test is just checking if the round-trip conversion works for all fields.

File: apis/run/v1alpha1/conversion.go (hunk 2 above)
Comment (ShashankGirish) [reply apis/run/v1alpha1/conversion.go:88]: Handled in cf38543c53922f7ac1aee651fbe9eb89284e4daa

File: Makefile [hunk 5]
```diff
@@ -624,6 +625,20 @@ generate-ui-swagger-api: ## Generate swagger files for UI backend
 	git checkout HEAD ${UI_DIR}/server/restapi/server.go
 	$(MAKE) fmt
 
+.PHONY: clean-generated-conversions
+clean-generated-conversions: ## Remove files generated by conversion-gen from the mentioned dirs. Example SRC_DIRS="./api/run/v1alpha1"
+	(IFS=','; for i in $(SRC_DIRS); do find $$i -type f -name 'zz_generated.conversion*' -exec rm -f {} \;; done)
+
+.PHONY: generate-go-conversions
+generate-go-conversions: $(CONVERSION_GEN) ## Generate conversions go code
+	$(MAKE) clean-generated-conversions SRC_DIRS="./apis/run/v1alpha1,./apis/run/v1alpha1"
```
[the remark this answers is not in the document]
Comment (ShashankGirish) [reply Makefile:634]: Great catch! 
We only need to run this in the spokes, so the second one is not necessary.

[the remark this answers is not in the document]
Comment (ShashankGirish) [reply Makefile]: Handled in 778c8d3ffd1f23b91230f7150533faef60e295de

**Review by imikushin — APPROVED** [review]

**Pull request merged**